File: Day 6/day6.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <cmath>

std::vector<int> times = {};
std::vector<int> distances = {};

bool isNumber(char c);
int processRace(int time, int distance);
int discriminant(int a, int b, int c);
bool hasSolution(int a, int b, int c);

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cout << "File name not provided\n";
        return 0;
    }

    std::string s;
    std::ifstream myFile;
    myFile.open(argv[1]);

    int i = -1;
    if (myFile.is_open()) {
        while (myFile.good()) {
            myFile >> s;
            std::cout << s << "\n";
            if (!isNumber(s[0])) {
                i++;
            } else {
                switch (i) {
                    case 0:
                        times.push_back(std::stoi(s));
                        break;
                    case 1:
                        std::cout << "Distance push: " << s << "\n";
                        distances.push_back(std::stoi(s));
                        break;
                }
            }
            if (myFile.eof()) {
                break;
            }
        }
    }
    int numRaces = times.size();
    int finalOutput = 0;
    for (int i = 0; i < numRaces; i++) {
        int processedNumber = processRace(times[i], distances[i]);
        std::cout << "Processed number: " << processedNumber << "\n";
        if (processedNumber > 0) {
            if (finalOutput == 0)
                finalOutput = processedNumber;
            else
                finalOutput *= processedNumber;
        }
    }
    std::cout << finalOutput << "\n";
}

bool isNumber(char c) {
    return c >= 48 and c <= 57;
}

int processRace(int time, int distance) {
    std::cout << "Discriminant: " << discriminant(1, -time, distance) << "\n";
    if (hasSolution(1, -time, distance)) {
        float solution1 = (-(-time) + sqrt(discriminant(1, -time, distance))) / 2.0;
        float solution2 = (-(-time) - sqrt(discriminant(1, -time, distance))) / 2.0;
        std::cout << "Solutions: " << solution1 << ", " << solution2 << "\n";
        if (solution2 > solution1) {
            return std::floor(solution2) - std::max(int(std::ceil(solution1)), 0) + 1;
        } else {
            return std::floor(solution1) - std::max(int(std::ceil(solution2)), 0) + 1;
        }
    } else {
        return 0;
    }
}

int discriminant(int a, int b, int c) {
    return b * b - 4 * a * c;
}

bool hasSolution(int a, int b, int c) {
    return discriminant(a, b, c) >= 0;
}

File: Day 6/day6.py
import math

fileName = "day6.in"

# inequality: (time - x) * x >= distance
# -x^2 + time * x - distance >= 0
# x^2 - time * x + distance <= 0
def calculateSolution(a, b, c):
    solution1 = (-b + math.sqrt(b * b - 4 * a * c)) / (2 * a)
    solution2 = (-b - math.sqrt(b * b - 4 * a * c)) / (2 * a)
    return solution1, solution2

def findNumberOfWays(time, distance):
    solution1, solution2 = calculateSolution(1, -time, distance)
    if (solution1 > solution2):
        return math.floor(solution1) - math.ceil(max(solution2, 0)) + 1
    else:
        return math.floor(solution2) - math.ceil(max(solution1, 0)) + 1

with open(fileName, 'r') as f:
    index: int = 0
    times = []
    distances = []
    for line in f:
        sep = line.split(' ')[1:]
        filtered = filter(lambda x : x, sep)
        mapped = map(lambda x : int(x), filtered)
        if index == 0:
            times = list(mapped)
            index += 1
        else:
            distances = list(mapped)
    finalOutput: int = 0
    for i in range(len(times)):
        if (finalOutput == 0):
            finalOutput = findNumberOfWays(times[i], distances[i])
        else:
            finalOutput *= findNumberOfWays(times[i], distances[i])
    print(finalOutput)